src/Compilers/CSharp/Portable/Parser/LexerCache.cs

```diff
@@ -30,25 +28,73 @@ internal partial class LexerCache
                                 return kind;
                             });
 
-        private readonly TextKeyedCache<SyntaxTrivia> _triviaMap;
-        private readonly TextKeyedCache<SyntaxToken> _tokenMap;
-        private readonly CachingIdentityFactory<string, SyntaxKind> _keywordKindMap;
+        private TextKeyedCache<SyntaxTrivia> _triviaMap;
+        private TextKeyedCache<SyntaxToken> _tokenMap;
+        private CachingIdentityFactory<string, SyntaxKind> _keywordKindMap;
         internal const int MaxKeywordLength = 10;
 
+        private PooledStringBuilder _stringBuilder;
+        private readonly char[] _identBuffer;
+        private SyntaxListBuilder _leadingTriviaCache;
+        private SyntaxListBuilder _trailingTriviaCache;
+
+        private const int LeadingTriviaCacheInitialCapacity = 128;
+        private const int TrailingTriviaCacheInitialCapacity = 16;
+
         internal LexerCache()
         {
             _triviaMap = TextKeyedCache<SyntaxTrivia>.GetInstance();
             _tokenMap = TextKeyedCache<SyntaxToken>.GetInstance();
             _keywordKindMap = s_keywordKindPool.Allocate();
+
+            _stringBuilder = PooledStringBuilder.GetInstance();
+            _identBuffer = new char[32];
+            _leadingTriviaCache = new SyntaxListBuilder(LeadingTriviaCacheInitialCapacity);
+            _trailingTriviaCache = new SyntaxListBuilder(TrailingTriviaCacheInitialCapacity);
         }
 
         internal void Free()
         {
             _keywordKindMap.Free();
+            _keywordKindMap = s_keywordKindPool.Allocate();
```

Comment:
>_keywordKindMap = s_keywordKindPool.Allocate(); [](http://example.com/codeflow?start=12&length=47)

It feels counterintuitive that a `Free` method doesn't actually free, but allocates. If we need a clear method, let's add it or rename this one if we would like to change its purpose.